app/gui/view/graph-editor/src/free_place_finder.rs

```diff
@@ -0,0 +1,255 @@
+//! The module containing an algorithm for searching free space for some point going at the specific
+//! direction.
+//!
+//! This is used in Graph Editor to find unoccupied place for newly created node.
+
+use crate::prelude::*;
+
+use ordered_float::OrderedFloat;
+
+
+
+// ====================
+// === OccupiedArea ===
+// ===================
+
+/// The structure describing an occupied area.
+///
+/// All such areas are rectangles described by x and y ranges.
+#[allow(missing_docs)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct OccupiedArea {
+    pub x1: f32,
+    pub y1: f32,
+    pub x2: f32,
+    pub y2: f32,
+}
+
+impl OccupiedArea {
+    /// Get x position of the left boundary.
+    pub fn left(&self) -> f32 {
+        min(self.x1, self.x2)
+    }
+
+    /// Get x position of the right boundary.
+    pub fn right(&self) -> f32 {
+        max(self.x1, self.x2)
+    }
+
+    /// Get y position of the top boundary.
+    pub fn top(&self) -> f32 {
+        max(self.y1, self.y2)
+    }
+
+    /// Get y position of the bottom boundary.
+    pub fn bottom(&self) -> f32 {
+        min(self.y1, self.y2)
+    }
+
+    /// Check if the rectangle contains given point.
+    ///
+    /// The boundaries are open - they are not considered occupied.
+    pub fn contains(&self, point: Vector2) -> bool {
+        (self.x1 - point.x) * (self.x2 - point.x) < 0.0
+            && (self.y1 - point.y) * (self.y2 - point.y) < 0.0
+    }
+
+    /// Return the x position of the left or right boundary, depending on whether the direction
+    /// points leftwards of rightwards respectively. Returns `None` if direction does not point
```

Comment:
In Rust the doc syntax for mentioning things is ``` [`None`] ```, not ``` `None` ``` (unfortunately). Please apply in other comments too.